### DISCUSSION TITLE
Installing OpenMage with Composer

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#4088] [Questions & Answers]
I have never installed OpenMage with Composer, so I followed the steps I found on this page

1. I created an openmage/ directory and accessed it. This is the root directory of the project, when I access the shop URL in browser
2. I ran the ```composer init``` command and answered all the questions
3. I followed the steps for composer configuration
```
# Allow composer to apply patches to dependencies of magento-lts
composer config --json extra.enable-patching true

# Configure Magento core composer installer to use magento-lts as the Magento source package
composer config extra.magento-core-package-type magento-source

# Configure the root directory that magento-lts will be installed to, such as "pub", "htdocs", or "www"
composer config extra.magento-root-dir pub
```
4. I followed the step for PHP 8 requirements

```
composer require "aydin-hassan/magento-core-composer-installer":"^2.1.0"
```

5. I installed the latest version

```
composer require "openmage/magento-lts":"^20.0.0"
```

I was asked about certain packages if they are safe and I completed successfully. Listing the openmage/ directory I am getting the following directories and files

```ssh
drwxr-xr-x  3 ubuntu ubuntu 4.0K Jul  7 14:51 app
-rw-r--r--  1 ubuntu ubuntu  785 Jul  7 15:07 composer.json
-rw-r--r--  1 ubuntu ubuntu  87K Jul  7 15:07 composer.lock
drwxr-xr-x 12 ubuntu ubuntu 4.0K Jul  7 15:07 pub
drwxr-xr-x  2 ubuntu ubuntu 4.0K Jul  7 14:53 src
drwxr-xr-x 20 ubuntu ubuntu 4.0K Jul  7 15:07 vendor
```

If I will access the store in browser obviously I will get the Forbidden page, because the openmage/ directory is the Apache root.

**Question 1**
Do I need to create a redirect to the OpenMage files in an .htaccess file?

**Question 2**
Which directory should I redirect to, **pub/** or **vendor/openmage/magento-lts/**? I found that both directories take 83 MB of space. 

**Question 3**
Why there are two OpenMage directories? When the composer updates it will update  the vendor/openmage/magento-lts directory not the pub/. Why the pub/ directory is not a symlink to the vendor/openmage/magento-lts?

**Comment by m-overlund** [comment]
1. I created a project composer.json in public_html - with OpenMage installed into a pub/public subfolder, and changed the webserver docroot folder to this.
2. you should use `pub` as docroot. You can see composer strategies (symlink vs. copy) here: https://github.com/OpenMage/magento-lts/blob/main/composer.json#L94
3. Someone properly can answer this - i can't remember the reason, but there is one.

**Reply by addison74** [reply]
**Related to question 1**. It was my mistake. The docroot directive is only used in the Apache configuration. If I choose a redirection, this can be done in .htaccess, but the URL will get the /pub directory in the string, which is to be avoided.

**Comment by Flyingmana** [comment]
for question 2 and 3

The one in pub/ is the correct one, as modules get installed into this one.
using symlinks should be possible via a config, but there are often various issues, especially if people try to change some of the files or want to deploy the code via ftp, which still some people with less developer experience prefer.

**Comment by addison74** [comment]
Thank you both for your answers. If I run OpenMage from **/pub**, then I can no longer use Composer to update it, as long as it affects the files from **/vendor/openmage/magento-lts**. However, why do we get two identical directories as content after installation?

**Reply by fballiano** [reply]
running openmage from /pub just means setting the docroot of the web server to the /pub directory, you don't need to copy anything so you will still update OM with composer as usual.

there are 2 directories because the M1 autoloader doesn't support loading from the vendor/openmage/etc folder but it needs (for historic reasons) to be in the root. that's why every composer based installation has to use the composer-magento plugin which is the one that copies all OM files from the vendor folder (where composer puts them) to the root, to make them work.

**Reply by addison74** [reply]
I am starting to understand the reasoning. Basically **/vendor/openmage/magento-lts** is a temporary directory. When Composer updates OpenMage, it first brings the files here and then copies them to /pub. With the disk space that is now offered for a VPS/VDS, it is not a problem to keep 100 MB permanently allocated.

**Reply by fballiano** [reply]
it kinda is yes, but it's because all the M1 autoload was conceived before composer even existed and it doesn't work the way composer would want to

**Reply by luigifab** [reply]
to save disk space you can remove content of vendor/openmage/magento-lts/* - then create an empty file inside, for example: vendor/openmage/magento-lts/keep-me-empty.txt
